Taproot wallet support

instagibbs opened the pull request:
This switches CLN to issuing p2tr addresses for deposits optionally, and also by default sending change to p2tr outputs.

Open questions:
1) should we deprecate wrapped segwit first?
2) I'm not happy with how many places in the code I have to check things like `chainparams->is_elements`(12 times looks like), bet it would be good to have some simpler way to not forget what version we're doing?

instagibbs commented:
one more libwally update to go....

instagibbs commented on wallet/walletrpc.c:
```diff
@@ -68,6 +78,27 @@ encode_pubkey_to_addr(const tal_t *ctx,
+					   (u8 *)x_key, 32,
+					   0);
+
+		/* Would be better to use wally_scriptpubkey_to_address, but lacks signet support */
```
will revisit this...

rustyrussell commented on bitcoin/script.h:
```diff
@@ -153,9 +166,15 @@ bool is_p2wsh(const u8 *script, struct sha256 *addr);
 
-/* Is this one of the four above script types? */
+/* Is this a taproot output? (extract xonly_pubkey bytes if not NULL) */
+bool is_p2tr(const u8 *script, u8 *xonly_pubkey);
```
Nitpick: as a hint, can we declare this as `bool is_p2tr(const u8 *script, u8 xonly_pubkey[32]);` ?  It's the same, but makes it clear that it should be 32 bytes long...

instagibbs replied:
done

rustyrussell commented on common/psbt_keypath.c:
```diff
@@ -14,20 +13,30 @@ void psbt_set_keypath(u32 index, const struct ext_key *ext, struct wally_map *ma
-		abort();
+	if (is_taproot) {
+		if (wally_psbt_output_taproot_keypath_add(output,
+												 ext->pub_key + 1, sizeof(ext->pub_key) - 1,
+												 NULL, 0,
+												 fingerprint, sizeof(fingerprint),
+												 path, 1) != WALLY_OK)
```
Guessing your local tabs are 4 spaces?  Because this is very indented!

instagibbs replied:
ok im in libwally mode; what's the expected indent level :)

rustyrussell commented on plugins/spender/multifundchannel.c:
```diff
@@ -1504,16 +1504,6 @@ perform_fundpsbt(struct multifundchannel_command *mfc, u32 feerate)
 				tal_fmt(tmpctx, "%zu", startweight));
 	}
 
-	/* If we've got v2 opens, we need to use a min weight of 110. */
-	/* BOLT-f53ca2301232db780843e894f55d95d512f297f9 #3:
-	 * The minimum witness weight for an input is 110.
-	 */
-	if (dest_count(mfc, OPEN_CHANNEL) > 0) {
-		json_add_string(req->js, "min_witness_weight",
-				tal_fmt(tmpctx, "%u", 110));
-	}
-
-
```
I don't think you can just delete this?

instagibbs replied:
ok this one is obviously true as this is an *empty* input, putting back

rustyrussell reviewed (commented):
One minor weird delete...

It's unfortunate that we can't simply reuse "bech32" as the JSON field name for these addresses: I guess some software won't be up-to-date.

However, I think it's naive to use "p2tr" as well, since we're hoping that THIS TIME FOR SURE we have a universal address format: next change it would be great not to have everyone have to parse a new field.  So, do we optimistically go for "bech32m"?

I may remove p2sh as part of the deprecations PR, let's not get bogged down behind that!

instagibbs commented:
> However, I think it's naive to use "p2tr" as well, since we're hoping that THIS TIME FOR SURE we have a universal address format: next change it would be great not to have everyone have to parse a new field. So, do we optimistically go for "bech32m"?

If someday there's witness output v2 that also uses bech32m, we'll have to un-camp it? p2tr just seems more explicit to me.

rustyrussell commented:
> last question I'd have: is it better to not switch to taproot change outputs, just to limit blast radius if something were to go wrong, at least for a release?

Unfortunately, I don't think we'd find any bugs without this, and it adds another step change.

However, I do note that you need to add two Changelog-Added lines, and edit the newaddr man page (which doesn't mention `p2tr` as an option for `addresstype`:

Changelog-Added: JSON-RPC: `newaddr`: `p2tr` option to create taproot addresses.
Changelog-Changed: Wallet: we now use taproot change addresses.

> pushed a couple fixes to tests as well
> 
> edit: can't replicate `tests/test_connection.py::test_feerate_stress` failures locally

That's a known flake, ignore...

rustyrussell commented:
Also, please change all `v23.05` to `v23.08`!

rustyrussell reviewed (changes requested):
See comments...

instagibbs commented:
> Also, please change all v23.05 to v23.08!

done!

also added description for arg in `doc/lightning-newaddr.7.md`

instagibbs commented:
rebased on the now-ACK'd https://github.com/ElementsProject/lightning/pull/6308 

removed the random gating in tests, things should be passing now...

rustyrussell commented:
Rebased on master, and folded the fixups into a single commit.  Also did some indent fixups.

niftynei commented:
Ayyyy!!!

On Mon, Jul 10, 2023 at 21:11 Rusty Russell ***@***.***>
wrote:

> Merged #6035 <https://github.com/ElementsProject/lightning/pull/6035>
> into master.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ElementsProject/lightning/pull/6035#event-9781729258>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIMAKN2GKIZWOMAPDOVWPDXPSY47ANCNFSM6AAAAAAVHHMQ3Q>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

harding commented:
Regarding [field names](https://github.com/ElementsProject/lightning/pull/6035#issuecomment-1562690919), @sipa previously had this advice for me:

```
<sipa> harding: i think output types should be distinguished based on sender's abilities to
    pay them, as you may need to construct different addresses for older/new sender software
<sipa> and in that regard, i think legacy/p2sh/bech32, and unfortunately, bech32m are
   probably the right choices
[...]
<harding> sipa: you're saying that (in December 2021) if I run `getnewaddress bech32m` I
    should get a P2TR address, and (years from now) if I run `getnewaddress bech32m` I should
    get a P2FOO address because all I care about is whether or not the person who will be spending
    to me can support bech32m?
<sipa> yes
```

Source: https://www.erisian.com.au/bitcoin-core-dev/log-2021-06-28.html#l-89
